### PR TITLE
[Function](array) support array_difference function

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_difference.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+mysql> select *,array_difference(k2) from array_type_table;
++------+-----------------------------+---------------------------------+
+| k1   | k2                          | array_difference(`k2`)          |
++------+-----------------------------+---------------------------------+
+|    0 | []                          | []                              |
+|    1 | [NULL]                      | [NULL]                          |
+|    2 | [1, 2, 3]                   | [0, 1, 1]                       |
+|    3 | [1, NULL, 3]                | [0, NULL, 3]                    |
+|    4 | [0, 1, 2, 3, NULL, 4, 6]    | [0, 1, 1, 1, NULL, 4, 2]        |
```
Comment (HappenLee) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_difference.md:58]: 4 - null = null

File: be/src/vec/functions/array/function_array_difference.h [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                        static_cast<Result>(curr_element) - static_cast<Result>(prev_element);
+                prev_element = curr_element;
+            }
+        }
+    }
+
+    template <typename Element, typename Result>
+    ColumnPtr _execute_number_expanded(const ColumnArray::Offsets64& offsets,
+                                       const IColumn& nested_column, ColumnPtr nested_null_map) {
+        using ColVecType = ColumnVectorOrDecimal<Element>;
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_difference.h:106]: the `Result` type should be deduce by `Element`

File: be/src/vec/functions/array/function_array_difference.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        } else if (which.is_uint32() || which.is_uint64() || which.is_int32()) {
+            return_type = std::make_shared<DataTypeInt64>();
+        } else if (which.is_int64() || which.is_int128()) {
+            return_type = std::make_shared<DataTypeInt128>();
+        } else if (which.is_float32() || which.is_float64()) {
+            return_type = std::make_shared<DataTypeFloat64>();
+        } else if (which.is_decimal()) {
+            return arguments[0];
+        }
+        return std::make_shared<DataTypeArray>(is_nullable ? make_nullable(return_type)
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_difference.h:70]: if return_type is nullptr, we should just return nullptr

File: be/src/vec/functions/array/function_array_difference.h [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+private:
+    template <typename Element, typename Result>
+    static void impl(const Element* __restrict src, Result* __restrict dst, size_t begin,
+                     size_t end) {
+        size_t curr_pos = begin;
+        if (curr_pos < end) {
+            Element prev_element = src[curr_pos];
+            dst[curr_pos] = {};
+            curr_pos++;
+            Element curr_element = src[curr_pos];
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_difference.h:97]:
```suggestion
```

File: be/src/vec/functions/array/function_array_difference.h [hunk 5]
```diff
@@ -67,8 +67,12 @@ class FunctionArrayDifference : public IFunction {
         } else if (which.is_decimal()) {
             return arguments[0];
         }
-        return std::make_shared<DataTypeArray>(is_nullable ? make_nullable(return_type)
-                                                           : return_type);
+        if (return_type) {
+            return std::make_shared<DataTypeArray>(is_nullable ? make_nullable(return_type)
+                                                               : return_type);
+        } else {
+            return nullptr;
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_difference.h:74]: maybe need DCHECK or LOG_FATAL

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**